src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1068 @@
+      owner_map[owner].end_time = sp.completion_time;
+    }
+    else {
+      owner_map[owner].start_time = sp.creation_time;
+      owner_map[owner].aggregated_time = duration;
+      owner_map[owner].end_time = sp.completion_time;
+    }
+    ldpp_dout(dpp, 10) << __func__ << "::Got " << name
+		       << " stats for shard #" << shard << dendl;
+  }
+
+  //---------------------------------------------------------------------------
+  static void show_dedup_ratio(const worker_stats_t &wrk_stats_sum,
+			       const md5_stats_t    &md5_stats_sum)
+  {
+    uint64_t rados_bytes_before = md5_stats_sum.rados_bytes_before_dedup;
+    uint64_t s3_bytes_before    = wrk_stats_sum.ingress_obj_bytes;
+    uint64_t s3_dedup_bytes     = md5_stats_sum.duplicated_blocks_bytes;
+    uint64_t s3_bytes_after     = s3_bytes_before - s3_dedup_bytes;
+    // skipped objects should be accounted for
+    // TBD: double check the logic!
+    uint64_t skipped_bytes = (wrk_stats_sum.ingress_skip_too_small_bytes +
+			      wrk_stats_sum.non_default_storage_class_objs_bytes);
+    s3_bytes_after -= skipped_bytes;
+    if (rados_bytes_before > s3_bytes_after && s3_bytes_after) {
+      double dedup_ratio = (double)rados_bytes_before/s3_bytes_after;
+      std::cout << "rados_bytes_before = " << rados_bytes_before << "\n";
+      std::cout << "s3_bytes_before    = " << s3_bytes_before << "\n";
+      std::cout << "s3_bytes_after     = " << s3_bytes_after << "\n";
+      std::cout << "dedup_ratio        = " << dedup_ratio << "\n";
+    }
+  }
+
+  //---------------------------------------------------------------------------
+  static void display_progress(uint64_t progress_a, uint64_t progress_b, unsigned shard)
+  {
+    if (progress_a == SP_NO_OBJECTS && progress_b == SP_NO_OBJECTS) {
+      return;
+    }
+
+    std::cout << std::hex << "0x" << std::setw(2) << std::setfill('0') << shard << std::dec << "] ";
+    if (progress_b == SP_ALL_OBJECTS) {
+      std::cout << "Token is marked completed! obj_count=" << progress_a << std::endl;
+    }
+    else if (progress_a == SP_NO_OBJECTS && progress_b == SP_NO_OBJECTS) {
+      std::cout << "Token was not started yet!" << std::endl;
+    }
+    else if (progress_b == SP_NO_OBJECTS) {
+      std::cout << "Token is incomplete: progress=" << progress_a << std::endl;
+    }
+    else {
+      std::cout << "Token is incomplete: progress = [" << progress_a
+		<< ", " << progress_b << "]" <<  std::endl;
+    }
+  }
+
+  //---------------------------------------------------------------------------
+  // command-line called from radosgw-admin.cc
+  int cluster::collect_all_shard_stats(rgw::sal::RadosStore *store,
+				       const DoutPrefixProvider *dpp)
+  {
+    librados::IoCtx ioctx;
+    if (init_dedup_pool_ioctx(store->getRados(), dpp, ioctx) != 0) {
+      return -1;
+    }
+
+    dedup_epoch_t epoch;
+    if (get_epoch(ioctx, dpp, &epoch, nullptr) != 0) {
+      return -1;
```

Comment:
why not reply with the `get_epoch()` return code?

Reply:
done